onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -330,17 +370,38 @@ Status CUDAExecutionProvider::OnRunStart() {
   auto& current_deferred_release_event = GetPerThreadContext().GetCurrentDeferredReleaseEvent();
   CUDA_RETURN_IF_ERROR(cudaEventCreate(&current_deferred_release_event, cudaEventDisableTiming));
   deferred_release_cpu_ptr_.emplace(current_deferred_release_event, DeferredReleaseCPUPtrs());
+
+  if (ConfiguredForGraphCapture() && !GetPerThreadContext().IsGraphCaptured()) {
+    if (GetPerThreadContext().FinishRegularRunsBeforeGraphCapture()) {
+      LOGS_DEFAULT(INFO) << "Capturing the CUDA Graph for this model";
+      GetPerThreadContext().CaptureBegin();
+    } else {
+      LOGS_DEFAULT(INFO) << "Waiting for the regular run before CUDA Graph capture to finish";
+    }
+  }
   return Status::OK();
 }
 
 Status CUDAExecutionProvider::OnRunEnd(bool sync_stream) {
+  if (ConfiguredForGraphCapture() && !GetPerThreadContext().IsGraphCaptured()) {
+    if (GetPerThreadContext().FinishRegularRunsBeforeGraphCapture()) {
+      GetPerThreadContext().CaptureEnd();
+      // CUDA work issued to a capturing stream doesn’t actually run on the GPU,
+      // so run the captured graph here to actually execute the work.
+      ORT_RETURN_IF_ERROR(GetPerThreadContext().GraphReplay());
+    } else {
+      GetPerThreadContext().IncrementRegularRunCountBeforeGraphCapture();
+    }
+  }
   // record deferred release event on default stream, and release per_thread_context
   auto current_deferred_release_event = GetPerThreadContext().GetCurrentDeferredReleaseEvent();
   CUDA_RETURN_IF_ERROR(cudaEventRecord(current_deferred_release_event, static_cast<cudaStream_t>(GetComputeStream())));
   if (sync_stream) {
     CUDA_RETURN_IF_ERROR(cudaStreamSynchronize(static_cast<cudaStream_t>(GetComputeStream())));
   }
-  ReleasePerThreadContext();
+  if (!ConfiguredForGraphCapture()) {
+    ReleasePerThreadContext();
```

Comment:
The below comment is added:
```
  // If cuda graph is enabled, the per thread context will not be released
  // because the per thread cuda graph needs to be maintained and replayed for
  // the next run.
```